### PR TITLE
diagonal check in commutation analysis

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
Attempts to speed up commutation analysis pass by avoiding matrix multiplication between two gates if the two gates are diagonal, which always commute.

It may be possible to obtain additional improvement if gates were to store a `_diagonal` property or transpiler directive.

### CODE REVIEW

File: qiskit/transpiler/passes/optimization/commutation_analysis.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ def run(self, dag):
 
 
 def _commute(node1, node2, cache):
+    from qiskit.quantum_info.operators.predicates import is_diagonal_matrix
```
Comment (mtreinish) [inline qiskit/transpiler/passes/optimization/commutation_analysis.py:90]: Is there a reason to do this at runtime (like a circular dependency)? If not can we do this at the top so it's done at import time.

File: qiskit/transpiler/passes/optimization/commutation_analysis.py [hunk 2]
```diff
@@ -111,15 +112,21 @@ def _commute(node1, node2, cache):
 
     node1_key = (node1.op.name, str(node1.op.params), str(qarg1))
     node2_key = (node2.op.name, str(node2.op.params), str(qarg2))
+
+    op1 = Operator(node1.op)
```
Comment (mtreinish) [inline qiskit/transpiler/passes/optimization/commutation_analysis.py:116]: Do we actually need the Operator class here or can we just work with the raw ndarray? There can often be a speedup if we work directly with numpy directly

File: qiskit/transpiler/passes/optimization/commutation_analysis.py [hunk 3]
```diff
@@ -111,15 +112,21 @@ def _commute(node1, node2, cache):
...
+    op2 = Operator(node2.op)
+    if is_diagonal_matrix(op1.data) and is_diagonal_matrix(op2.data):
+        # may be better for short depth wide circuits; use depth criteria?
+        return True
```
Comment (mtreinish) [inline qiskit/transpiler/passes/optimization/commutation_analysis.py:120]: Does it make sense to cache this too? I'm a bit worried that for circuits with identical gates we'll end up spending a lot of time checking if a matrix is diagonal

**Comment by ewinston** [comment]
I tried timing the speed of the pass when handling circuits with random 2-qubit gates, some fraction of which are diagonal. Running 10 cycles of 1000 gates at each fraction attempted seemed to be slower than what's currently implemented.
This version:

image

Which is slower not doing the diagonal check;

image

In the special case that the diagonal gates are all the same diagonal gate with ~80% population then the diagonal check is a little faster. 
image

Closing this since the speedup seems unlikely for most circuits.

**Pull request closed without merging**